phone-book/src/app.js

```diff
@@ -0,0 +1,177 @@
+class App {
+  constructor(appContainerId) {
+    this.state = {};
+    this.appContainer = document.getElementById(appContainerId);
+    this.router = new Router(this);
+    this.serverAPI = new ServerAPI();
+    const userPage = new EditUser(this);
+    this.pages = {
+      contacts : {
+        pageObject : new Contacts(this),
+        href : '/contacts.html'
+      },
+      keypad : {
+        pageObject : new Keypad(this),
+        href : '/keypad.html'
+      },
+      editUser : {
+        pageObject : userPage,
+        href : '/edit-contact.html'
+      },
+      user : {
+        pageObject : new User(this),
+        href : '/user.html'
+      },
+      addUser : {
+        pageObject : userPage,
+        href : '/add-user.html'
+      }
+    }
+
+    const appHtml = this.renderAppContainer();
+    this.appContainer.innerHTML = appHtml;
+    this.addRouter();
+
+    this.state.pageName = 'Contacts';
+    this.updateState({activePage : 'contacts'});
+    this.router.gotToPage(this.state,'/index.html');
+    this.changePageToActive();
+  }
+
+  updateState(newState){
+    this.state = {};
+    Object.assign(this.state,newState);
+
+
+  }
+
+  addRouter(){
+    this.appContainer.addEventListener('click', (event) => {
+      event.preventDefault();
+      const container = event.currentTarget;
+      let target = event.target;
+      while (target != container) {
```

Comment:
I don't like such imperative loop, you could be sure about the target and don't even have any conditionals inside